walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,126 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public static final int DAY_START = TimeRange.START_OF_DAY;
+    public static final int DAY_END = TimeRange.END_OF_DAY;
+    public static final TimeRange WHOLE_DAY = TimeRange.WHOLE_DAY;
+
+    /**
+      * Return a collection of TimeRange that attendees can meet for a requested 
+      * meeting. If there is not enough room for optional attendees, find TimeRanges 
+      * that just work for required attendees. If there are no required attendees, 
+      * find the TimeRangs for just the optional attendees.
+      */
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        long meetingDuration = request.getDuration();
+        // Required attendees for the request meeting.
+        Collection<String> requiredAttendees = request.getAttendees(); 
+        // All attendees (required and optional) for the request meeting.
+        Collection<String> allAttendees = new ArrayList<>(requiredAttendees);
+        allAttendees.addAll(request.getOptionalAttendees());
+
+        // If the meeting duration is invalid, return an empty Collection.
+        if (meetingDuration > WHOLE_DAY.duration() || meetingDuration <= 0) {
+            return Arrays.asList();
+        }
+
+        // Find available TimeRange for both required and optional attendees.
+        Collection<TimeRange> allMeetingTimes = 
+            findMeetingTimes(events, allAttendees, meetingDuration);
+
+        // If there is meeting time for all, return allMeetingTimes so that 
+        // both required and optional can attend the meeting. 
+        if (allMeetingTimes.size() > 0) {
+            return allMeetingTimes;
+        }
+
+        // If there is no required attendee, return allMeetingTimes for 
+        // optional attendees.
+        if (requiredAttendees.isEmpty()) {
+            return allMeetingTimes;
```

Comment:
Isn't the `allMeetingTimes` empty at this point? I think it make sense to either move this before the if from line 53 or merge the condition inside that if. This is also a topic worth discussion in the daily sync. :-)

Reply:
I asked Sue to move this condition out of the if in Line53 since it's easier to read for me. While it might be easier to understand if just return an empty list.

Reply:
It is not necessarily an empty list. `allMeetingTimes` would contain some TimeRanges for optional attendees, even if `requiredAttendees` is an empty collection. While it might be an empty collection, but we still need to use `findMeetingTimes` function to determine whether it is empty or not. :)

Reply:
Okay you are right! I'll just return an empty Collection. :)